pkg/grpcserver/creds_injector.go

```diff
@@ -0,0 +1,71 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+var minsBeforeExpiration = time.Minute * 5
+
+// CredsInjector implements credentials.PerRPCCredentials interface
+type CredsInjector struct {
+	tokenGenerator  func() (string, error)
+	m               sync.Mutex
+	currentToken    string
+	currentTokenExp int64
+	tlsEnabled      bool
+}
+
+// GetRequestMetadata checks JWT token expiration time and invokes token generator function to get new token if that is needed
+func (i *CredsInjector) GetRequestMetadata(context.Context, ...string) (map[string]string, error) {
+	i.m.Lock()
+	defer i.m.Unlock()
+	inAFewMins := time.Now().Unix() + int64(minsBeforeExpiration.Seconds())
+	if i.currentToken == "" || i.currentTokenExp == 0 || inAFewMins >= i.currentTokenExp {
+		token, err := i.tokenGenerator()
+		if err != nil {
+			return nil, fmt.Errorf("cannot generate token, err: %s", err.Error())
+		}
+		i.currentToken = token
+
+		t, _, err := new(jwt.Parser).ParseUnverified(i.currentToken, &jwt.StandardClaims{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse authorization token: %s", err.Error())
+		}
+
+		claims, ok := t.Claims.(*jwt.StandardClaims)
+		if !ok {
+			return nil, fmt.Errorf("failed to get token claims")
+		}
+
+		i.currentTokenExp = claims.ExpiresAt
+	}
+
+	return map[string]string{
+		"authorization": "Bearer " + i.currentToken,
+	}, nil
+}
+
+func (i *CredsInjector) RequireTransportSecurity() bool {
+	return i.tlsEnabled
+}
+
+// ResetToken makes CredsInjector to invoke token generation next time when GetRequestMetadata is invoked
+func (i *CredsInjector) ResetToken() {
+	i.m.Lock()
+	i.currentToken = ""
+	i.currentTokenExp = 0
+	i.m.Unlock()
+}
+
+// NewRPCCredentials creates gRPC interceptor to inject Authorization token in requests
+func NewRPCCredentials(generator func() (string, error), tlsEnabled bool) *CredsInjector {
```

Comment:
Code is correct, but normally NewXXX is the API where XXX is the name of the struct. Here NewRPCCredentials returns CredsInjector, which is confusing

Reply:
That makes sense, I just missed that while cleaning my PR :-(